### ISSUE TITLE
Activities fail a WCAG 2.1 AA Success Criterion when zoomed at 400%

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#1107]
Reported on forums by OU https://xerte.org.uk/index.php/en/forum/accessibility/2882-activities-fail-a-wcag-2-1-aa-success-criterion-when-zoomed-at-400
Xerte activities fail WCAG 2.1 AA Success Criterion (1.4.10: Reflow) when zoomed at 400%. The header and footer obscures the body content. Please view attachment. Please advise if this can be corrected.

I don't think the header is the problem but the footer is because it's position is absolute rather than moving with the content. I think this overlaps with another desirable change that some ask for where the footer is sticky but moves with the content if needed rather than being absolute. e.g. https://philipwalton.github.io/solved-by-flexbox/demos/sticky-footer/

So two issues I think:
Solving this issue with the current structure
Changing the structure (perhaps via an optional property) so it's possible to have a sticky footer rather than absolute footer.

I guess one of the problems might be that I think there are page types that currently calculate sizes and positions that might be affected by a change like this?

CSS tricks have what is perhaps a relevant article on different ways to implement a stick footer https://css-tricks.com/couple-takes-sticky-footer/

**Comment by FayCross** [comment]
@ronm123 yes, there are a handful of pages that use the x_footerBlock height in working out the space available to layout the page contents so these will need to be changed to cope with sticky footers.

At the moment, there are two layouts:
1. default - fixed header & footer with page content scrolling between them
2. mobile - fixed footer with header scrolling with the content above it (also reorganisation of contents on some pages e.g. columns)

Xenith determines when project loads which layout is used and this is fixed, even if the browser window is later reduced. It would be better for the layout to be more responsive and move between 1 & 2 depending on the browser size (like BS template is). Perhaps this is something that can also be looked at alongside the zooming issues

**Comment by FayCross** [comment]
Another related thing to look at is what the footer bar does when the buttons don't all fit on. I think the buttons on left should be minimised (with the >> button there to reveal them) if the zoom level means the footer bar goes like this:
![zoom](https://user-images.githubusercontent.com/8926470/172590804-9d93bc69-9666-4b13-ba27-dbe994e9449e.PNG)

**Comment by ronm123** [comment]
@FayCross yeah agree, what happens to the footer buttons should be more elegant than it currently is and also when viewed on a mobile the buttons are arguably too small but I think that was partially to allow all of the buttons to fit especially if some of the optional buttons are included like help and glossary etc.

Also agree that we shouldn't necessarily try to pick up zoom level and respond to that but not sure detecting % of the header and footer will resolve everything either. The main problem seems to be the fixed footer rather than header. If I view our page types demo on my iphone I can see that the header moves but the footer doesn't and if I use the aA option to zoom the footer gets so large that there is very little visible page left to see the content even though the header now moves. It would be better the other way around where the header is fixes but the footer moves down with the content. That would be desirable as an optional property anyway so that the author could choose a sticky rather than fixed footer option.

We could respond to the forum question to say we will look at solutions but is there already a solution they could implement with script/css etc?

**Comment by FayCross** [comment]
@ronm123 I'm not sure I agree about fixed header. Viewing the page examples project on my phone, the page with the longest title has the header block taking approx. 40% of height, while footer bar is maybe just 15% of height. Fixing header instead would make a lot less space for page content. Probably one to discuss at next dev meeting!

**Comment by ronm123** [comment]
@FayCross yeah sorry that comment was misleading I guess neither header or footer can be fixed. I was testing on a local copy of my accessibility testing LO and on a page with narration added. In that case the footer including the narration takes up more space but what I really imagined is that it needs to work more like the Flexbox example both on mobile and on desktop https://philipwalton.github.io/solved-by-flexbox/demos/sticky-footer/

We also need to be realistic here e.g. on mobile this is going to be mostly about reflowing and reading at the zoom level needed not for instance being able to complete drag and drop interactions on a mobile at that zoom level. I don't think we actually scheduled the next dev meeting did we?

How do we respond to the forum post? 
Is there anything that would help via script/css in the meantime?

**Comment by ronm123** [comment]
@FayCross do you mean you've been attending or delivering accessibility training?

Not sure this is something else rather than the same issue as reported by the OU just with a slightly different point of reference. You'll have seen that Alistair replied to the forum post so then I did too asking a couple of questions about how/why it was reported and also possible contribution. Not seen a reply yet though.

I think it's highly desirable to have at least an option for sticky header and footer but yes there's obviously going to be other challenges with that with different pages types either because of the sizing and layout or because some don't currently reflow into a single column. We should get another dev day in the diaries!

**Comment by JohnSmith-LT** [comment]
What about something like this as a temporary fix (the condition window.devicePixelRatio > 3 most likely needs to be more complex):

Create a next project, add a Plain Text page and add a Styles and Script box to that page.

Put this in the Styles box:
```
.fullHeight {
  height: 100% !important;
  margin: 0 !important;
}

.hideZoom {
  visibility: hidden;
}
```
And put this in the Script box:
```
window.addEventListener("resize", ()=>{
  if (window.devicePixelRatio >=4 ){
  	$("#x_headerBlock").addClass("hideZoom");
    $("#x_footerBlock").addClass("hideZoom");
     $("#x_pageHolder").addClass("fullHeight");
  }
  else {
  	$("#x_headerBlock").removeClass("hideZoom");
    $("#x_footerBlock").removeClass("hideZoom");
     $("#x_pageHolder").removeClass("fullHeight");
  }
});
```

Requires people to zoom back out to navigate etc but at least they can see the full content - not tested on other pages though....

Alternatively you can "unzoom" the header and footer with something like :
```
.unzoomHeader {
    transform: scale(0.5) !important;
    left: calc(-50% - 1px) !important;
    width: 200vw !important;
    top: -15px !important;
}
.unzoomFooter {
     transform: scale(0.5) !important;
    left: calc(-50% - 1px) !important;
    width: 200vw !important;
    bottom: -10px !important;
}
```
And:
```
window.addEventListener("resize", ()=>{
  if (window.devicePixelRatio >=4 ){
  	$("#x_headerBlock").addClass("unzoomHeader");
    $("#x_footerBlock").addClass("unzoomFooter");
  }
  else {
  	$("#x_headerBlock").removeClass("unzoomHeader");
    $("#x_footerBlock").removeClass("unzoomFooter");
  }
});
```
But it's harder to then deal with the content since the margin-top, margin bottom and height seem to be constantly recalculated...

Neither method is going to be perfect but is it better than just now?

**Comment by ronm123** [comment]
@JohnSmith-LT I just tested this with a local copy of the page types demo. I haven't gone through every page but it does seem like a reasonable fix and better than what happens right now. Not sure if we can more accurately pick up the PixelRatio for different resolution monitors? e.g. on my 4K monitor triggering it at >=6 is better than >=4 but on my had monitors >=4 seems about right.

I also wondered about adding an alert e.g.
`window.addEventListener("resize", ()=>{
  if (window.devicePixelRatio >=4 ){
  	$("#x_headerBlock").hide();
    $("#x_footerBlock").hide();
     $("#x_pageHolder").addClass("fullHeight");
    var alertshown='';
    if (alertshown != '1'){
    alert("Please note: at high zoom levels the footer and page navigation will automatically hide to allow you to scroll down and read all the contents. You will then need reduce the zoom level to access the footer and navigation again.");
    alertshown=1;
    }
    else{
      //do nothing
    }
  }
  else {
  	$("#x_headerBlock").show();
    $("#x_footerBlock").show();
     $("#x_pageHolder").removeClass("fullHeight");`

However I've probably got something wrong there because the alert is firing more than once and on the way up and on the way down.

**Comment by JohnSmith-LT** [comment]
Yeah the calculation is a lot more complex than just `window.devicePixelRatio >=4` because that also factors in pixel density, etc. On Retina displays it will be double so there may need to be some browser sniffing but I haven't come across any handy function that just works. If we can get the figure for some devices and the browser info then we can probably write something but it will never be foolproof or 100% perfect and may get broken with future updates...

The best way would be to use CSS3 rules to do it but would probably require substantial rewriting of the structure and CSS...

**Comment by FayCross** [comment]
This issue is noted on the 3.14 accessibility statement:

Zooming in XOT projects is effective to around 250%. Above this level, the space available to display the main content is drastically reduced due to the header and footer bars taking up a large part of the screen. This fails [1.4.10 (Reflow)](https://www.w3.org/WAI/WCAG22/quickref/#reflow).
